File: quiz_app_flutter/lib/presentation/result/ui/result_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:go_router/go_router.dart';
import 'package:quiz_app_flutter/presentation/result/model/result_screen_data.dart';
import 'package:quiz_app_flutter/presentation/result/notifiers/result_notifier.dart';
import 'package:quiz_app_flutter/presentation/utils/ui/widgets/circular_percentage_progress.dart';
import 'package:quiz_app_flutter/presentation/utils/ui/widgets/placeholder_scaffold.dart';
import 'package:quiz_app_flutter/presentation/utils/ui/widgets/quiz_app_top_appbar.dart';
import 'result_item_row.dart';

class ResultScreen extends ConsumerStatefulWidget {
  final String resultKey;

  const ResultScreen({super.key, required this.resultKey});

  @override
  ResultScreenState createState() => ResultScreenState();
}

class ResultScreenState extends ConsumerState<ResultScreen> with TickerProviderStateMixin {
  int selectedTabIndex = 0;
  late TabController _tabController;

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addPostFrameCallback((_) {
      ref.read(resultNotifierProvider.notifier).getResult(widget.resultKey);
    });
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    // Initialize TabController after dependencies are available
    final uiState = ref.watch(resultNotifierProvider);
    final tabCount = uiState.when(
      data: (data) {
        if (data == null) return 0;
        return [
          if (data.totalCorrectItems > 0) 1,
          if (data.totalSkippedItems > 0) 1,
          if (data.totalInCorrectItems > 0) 1,
        ].length;
      },
      loading: () => 0,
      error: (_, __) => 0,
    );
    _tabController = TabController(length: tabCount, vsync: this, initialIndex: selectedTabIndex);
    _tabController.addListener(() {
      if (_tabController.index != selectedTabIndex) {
        setState(() {
          selectedTabIndex = _tabController.index;
        });
      }
    });
  }

  @override
  void dispose() {
    _tabController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final uiState = ref.watch(resultNotifierProvider);

    return PlaceholderScaffold(
      toolbarConfig: QuizAppToolbar(
        title: uiState.when(
          data: (data) => data?.quizTitle ?? '',
          loading: () => 'Loading',
          error: (error, _) => 'Error',
        ),
        navigationIcon: Icons.arrow_back,
        navigationIconContentDescription: 'Back',
        onNavigationClick: () {
          context.pop();
        },
      ),
      asyncValue: uiState,
      onRetryClicked: () {
        ref.read(resultNotifierProvider.notifier).getResult(widget.resultKey);
      },
      bodyContent: (data) {
        final currentTabTitle = 'Current (${data?.totalCorrectItems})';
        final skippedTabTitle = 'Skipped (${data?.totalSkippedItems})';
        final incorrectTabTitle = 'Incorrect (${data?.totalInCorrectItems})';

        final tabs = [
          if (data != null && data.totalCorrectItems > 0) (currentTabTitle, data.correctItems),
          if (data != null && data.totalSkippedItems > 0) (skippedTabTitle, data.skippedItems),
          if (data != null && data.totalInCorrectItems > 0) (incorrectTabTitle, data.incorrectItems),
        ];

        return Container(
          padding: const EdgeInsets.all(16.0),
          child: CustomScrollView(
            slivers: [
              SliverToBoxAdapter(child: ResultReportCard(data: data!)),
              SliverToBoxAdapter(
                child: Column(
                  children: [
                    TabBar(controller: _tabController, tabs: tabs.map((tab) => Tab(text: tab.$1)).toList()),
                    const SizedBox(height: 16),
                  ],
                ),
              ),
              SliverList(
                delegate: SliverChildBuilderDelegate((context, index) {
                  final filteredItems = tabs.isNotEmpty ? tabs[selectedTabIndex].$2 : [];
                  final item = filteredItems[index];
                  return ResultItemRow(item: item);
                }, childCount: tabs.isNotEmpty ? tabs[selectedTabIndex].$2.length : 0),
              ),
            ],
          ),
        );
      },
    );
  }
}

class ResultReportCard extends StatelessWidget {
  final ResultScreenData data;

  const ResultReportCard({super.key, required this.data});

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        if (data.quizDescription != null)
          Padding(
            padding: const EdgeInsets.only(bottom: 16.0),
            child: Text(
              data.quizDescription!,
              style: Theme.of(context).textTheme.bodyLarge?.copyWith(fontWeight: FontWeight.bold),
            ),
          ),
        Card(
          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(12)),
          elevation: 4,
          child: Padding(
            padding: const EdgeInsets.all(16.0),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              crossAxisAlignment: CrossAxisAlignment.center,
              children: [
                Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Text('Total Questions: ${data.totalQuestions}'),
                    const SizedBox(height: 8),
                    Text('Incorrect Answers: ${data.totalInCorrectItems}'),
                  ],
                ),
                CircularPercentageProgress(
                  progress: data.resultPercentage / 100.0,
                  size: 48.0,
                  strokeWidth: 4.0,
                  progressColor: Colors.blue,
                  backgroundColor: Colors.grey.shade300,
                ),
              ],
            ),
          ),
        ),
      ],
    );
  }
}

File: quiz_app_kmp/composeApp/src/commonMain/kotlin/com/tnm/quizappkmp/presentation/result/ui/ResultScreen.kt
package com.tnm.quizappkmp.presentation.result.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material3.Card
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Tab
import androidx.compose.material3.TabRow
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.navigation.NavHostController
import com.tnm.quizappkmp.domain.model.result.ResultData
import com.tnm.quizappkmp.presentation.result.model.ResultScreenData
import com.tnm.quizappkmp.presentation.result.viewmodel.ResultViewModel
import com.tnm.quizappkmp.presentation.utils.state.QuizAppUiState
import com.tnm.quizappkmp.presentation.utils.ui.AppCardDefaults
import com.tnm.quizappkmp.presentation.utils.ui.AppColors
import com.tnm.quizappkmp.presentation.utils.ui.AppDimens
import com.tnm.quizappkmp.presentation.utils.ui.CircularPercentageProgress
import com.tnm.quizappkmp.presentation.utils.ui.PlaceholderScaffold
import com.tnm.quizappkmp.presentation.utils.ui.QuizAppToolbar
import com.tnm.quizappkmp.presentation.utils.ui.SpacerMediumHeight
import com.tnm.quizappkmp.presentation.utils.ui.toProgress
import org.jetbrains.compose.resources.stringResource
import org.koin.compose.viewmodel.koinViewModel
import quiz_master_kmp.composeapp.generated.resources.Res
import quiz_master_kmp.composeapp.generated.resources.label_back
import quiz_master_kmp.composeapp.generated.resources.label_error
import quiz_master_kmp.composeapp.generated.resources.label_loading
import quiz_master_kmp.composeapp.generated.resources.result_correct_answers
import quiz_master_kmp.composeapp.generated.resources.result_total_questions
import quiz_master_kmp.composeapp.generated.resources.tab_current
import quiz_master_kmp.composeapp.generated.resources.tab_incorrect
import quiz_master_kmp.composeapp.generated.resources.tab_skipped

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ResultScreen(
    key: String,
    navController: NavHostController
) {
    val viewModel: ResultViewModel = koinViewModel()

    val uiState by viewModel.state.collectAsState()
    var selectedTabIndex by remember { mutableIntStateOf(0) }

    LaunchedEffect(Unit) {
        viewModel.getResult(key)
    }
    PlaceholderScaffold(
        toolbarConfig = QuizAppToolbar(
            title = when (uiState) {
                QuizAppUiState.Loading -> stringResource(Res.string.label_loading)
                is QuizAppUiState.Success -> (uiState as QuizAppUiState.Success<ResultScreenData>).data.quizTitle.orEmpty()
                is QuizAppUiState.Error -> stringResource(Res.string.label_error)
            },
            navigationIcon = Icons.AutoMirrored.Filled.ArrowBack,
            navigationIconContentDescription = stringResource(Res.string.label_back),
            onNavigationClick = { navController.navigateUp() },
        ),
        uiState = uiState,
        modifier = Modifier
    ) { data ->
        // 1. Prepare filtered lists once
        val currentTabTitle = stringResource(Res.string.tab_current, data.totalCorrectItems)
        val skippedTabTitle = stringResource(Res.string.tab_skipped, data.totalSkippedItems)
        val incorrectTabTitle = stringResource(Res.string.tab_incorrect, data.totalInCorrectItems)

        // 2. Build dynamic tabs
        val tabs = remember(data.quizTitle) {
            mutableListOf<Pair<String, List<ResultData.Item>>>().apply {
                if (data.totalCorrectItems > 0) add(currentTabTitle to data.correctItems)
                if (data.totalSkippedItems > 0) add(skippedTabTitle to data.skippedItems)
                if (data.totalInCorrectItems > 0) add(incorrectTabTitle to data.incorrectItems)
            }
        }

        LazyColumn(
            modifier = Modifier
                .fillMaxSize()
                .padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp)
        ) {
            // Top: Summary Card as first item
            item {
                ResultReportCard(data)
            }
            // TabRow
            item {
                Column {
                    TabRow(selectedTabIndex = selectedTabIndex) {
                        tabs.forEachIndexed { index, tab ->
                            Tab(
                                selected = selectedTabIndex == index,
                                onClick = { selectedTabIndex = index },
                                text = { Text(tab.first) }
                            )
                        }
                    }
                    SpacerMediumHeight()
                }
            }
            // Filtered items for selected tab
            val filteredItems = tabs.getOrNull(selectedTabIndex)?.second ?: emptyList()

            // Add each item as a LazyColumn item with stable key
            items(
                items = filteredItems,
                key = { it.questionId } // use a stable unique ID if available
            ) { item ->
                ResultItemRow(item)
            }
        }
    }
}

@Composable
fun ResultReportCard(data: ResultScreenData) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(top = 16.dp)
    ) {
        data.quizDescription?.let {
            Text(
                text = it,
                fontWeight = FontWeight.Bold,
                style = MaterialTheme.typography.bodyLarge,
                modifier = Modifier.padding(bottom = 16.dp)
            )
        }

        Card(
            modifier = Modifier.fillMaxWidth(),
            shape = AppCardDefaults.shape,
            elevation = AppCardDefaults.elevation()
        ) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(16.dp),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.SpaceBetween
            ) {
                // Left side: Texts
                Column(
                    verticalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    Text(
                        text = stringResource(
                            Res.string.result_total_questions,
                            data.totalQuestions
                        )
                    )
                    Text(
                        text = stringResource(
                            Res.string.result_correct_answers,
                            data.totalCorrectItems
                        )
                    )
                }

                // Right side: Circular progress
                CircularPercentageProgress(
                    progress = data.resultPercentage.toProgress(),
                    size = AppDimens.ProgressIndicatorSize,
                    strokeWidth = AppDimens.ProgressStrokeWidth,
                    progressColor = AppColors.ProgressColor,
                    backgroundColor = AppColors.ProgressBackground
                )
            }
        }
    }// Column end
}